compiler/qsc_eval/src/intrinsic.rs

```diff
@@ -249,6 +250,82 @@ fn two_qubit_rotation(
     }
 }
 
+/// Performs relabeling of qubits from the a given left array to the corresponding right array.
+/// The function will swap qubits with the given function to match the new relabeling, returning an error
+/// if the qubits are not unique or if the relabeling is not a valid permutation.
+pub fn qubit_relabel(
+    arg: Value,
+    arg_span: PackageSpan,
+    mut swap: impl FnMut(usize, usize),
+) -> Result<Value, Error> {
+    let [left, right] = unwrap_tuple(arg);
+    let left = left
+        .unwrap_array()
+        .iter()
+        .map(|q| q.clone().unwrap_qubit().0)
+        .collect::<Vec<_>>();
+    let right = right
+        .unwrap_array()
+        .iter()
+        .map(|q| q.clone().unwrap_qubit().0)
+        .collect::<Vec<_>>();
+    let left_set = left.iter().collect::<FxHashSet<_>>();
+    let right_set = right.iter().collect::<FxHashSet<_>>();
+    if left.len() != left_set.len() || right.len() != right_set.len() {
+        return Err(Error::QubitUniqueness(arg_span));
+    }
+    if left_set != right_set {
+        return Err(Error::RelabelingMismatch(arg_span));
+    }
+
+    let mut map = FxHashMap::default();
```

Comment:
Would it make sense to initialize with `2*lhs_set.len()` elements? This would be an upper bound, just not sure which will be better for typical perf to avoid reallocations.

Reply:
Good idea! I think I can assert that the map will have exactly `left.len()` entries because if `left.len() != left_set.len()` or `left_set != right_set` we return an error above, so by this point I know exactly how many and which keys the map will have, just not whether any of them are getting skipped due to not being permuted (ie: the 1 in [0,1,2]->[2,1,0]). I'll add the reserve.

All that said, I expect the total size of this array to never get that big, since it scales based on the number of qubits being used.